pkg/lang/csharp/plugin_expose.go

```diff
@@ -0,0 +1,719 @@
+	types := FindDeclarationsInFile[*TypeDeclaration](file).Declarations()
+	controllers := filter.NewSimpleFilter(
+		predicate.AnyOf(
+			HasBaseWithSuffix("Controller"),
+			NameHasSuffix[*TypeDeclaration]("Controller"),
+			HasAttribute[*TypeDeclaration]("Microsoft.AspNetCore.Mvc.Controller"),
+			HasAttribute[*TypeDeclaration]("Microsoft.AspNetCore.Mvc.ApiController"),
+		),
+		predicate.Not(HasAttribute[*TypeDeclaration]("Microsoft.AspNetCore.Mvc.NonController")),
+	).Apply(types...)
+	var controllerSpecs []controllerSpec
+	for _, c := range controllers {
+		controller := *c
+		spec := controllerSpec{
+			name:                    strings.TrimSuffix(controller.Name, "Controller"),
+			class:                   controller,
+			controllerAttributeSpec: parseControllerAttributes(controller),
+			actions:                 findActionsInController(controller),
+			execUnitName:            h.Unit.Name,
+		}
+		controllerSpecs = append(controllerSpecs, spec)
+	}
+	return controllerSpecs
+}
+
+// resolveRoutes returns a list of gatewayRouteDefinitions
+// by merging a controller's annotations with those of its contained actions
+func (c controllerSpec) resolveRoutes() []gatewayRouteDefinition {
+	var routes []gatewayRouteDefinition
+
+	prefixes := c.routeTemplates
+	hasPrefixes := true
+	if len(c.routeTemplates) == 0 {
+		prefixes = []string{""}
+		hasPrefixes = false
+	}
+
+	for _, action := range c.actions {
+		for _, prefix := range prefixes {
+			routeTemplate := action.routeTemplate
+			// action route templates starting with "/" or "~/" indicate controller-level route prefixes should be ignored
+			if strings.HasPrefix(routeTemplate, "~/") {
+				routeTemplate = strings.TrimPrefix(routeTemplate, "~")
+			} else if !strings.HasPrefix(routeTemplate, "/") {
+				routeTemplate = path.Join("/", prefix, action.routeTemplate)
+			}
+
+			if !hasPrefixes && !action.hasRouteTemplate {
+				zap.L().Sugar().Debugf("%s cannot be mapped to a route: no route template found", action.method.QualifiedName)
+				continue
+			}
+
+			verb := action.verb
+			if verb == "" {
+				verb = core.VerbAny
+			}
+
+			// add fixed route for second to last segment if last routeTemplate segment is an optional param
+			// e.g. /a/{b?} -> /a, /a/:b
+			nonOptionalRoute := stripOptionalLastSegment(routeTemplate)
+			if nonOptionalRoute != routeTemplate {
+				routes = append(routes, gatewayRouteDefinition{
+					Route: core.Route{
+						Verb: verb,
+						Path: strings.ToLower(
+							sanitizeAttributeBasedPath(nonOptionalRoute, c.area, c.name, action.name)),
+						ExecUnitName:  c.execUnitName,
+						HandledInFile: c.class.DeclaringFile,
+					},
+					DefinedInPath: c.class.DeclaringFile,
+				})
+			}
+
+			routes = append(routes, gatewayRouteDefinition{
+				Route: core.Route{
+					Verb: verb,
+					// using lowercase routes enforces consistency
+					// since ASP.NET Core is case-insensitive and API Gateway is case-sensitive
```

Comment:
It can definitely be tricky. Are you referring to a compiler warning or documentation? I'm definitely in favor of documenting this restriction.

Reply:
We should probably warn if their path contains any capital letters. Also, since this is API Gateway specific this transformation should probably occur somewhere further down (runtime or IaC).

Reply:
I guess I can see bumping this down to the runtime level. Given it's a framework-level detail, I don't think moving this to IaC would make sense.

One of the main issues leading to this choice is that any token replacements would be pascal case by default. If there wasn't that quirk, it'd be much easier to commit to creating the routes as declared without touching their case.

Example:
```csharp
[Route("/api/[controller]")] // -> "/api/User"
public class UserController {}
```

Reply:
Looking into this as a followup PR.